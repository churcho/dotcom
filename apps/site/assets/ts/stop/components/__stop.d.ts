import { Stop } from "../../v3api";

export default interface StopPageData {
  stop: Stop;
}
