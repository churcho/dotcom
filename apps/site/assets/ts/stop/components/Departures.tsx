import React, { ReactElement } from "react";
import { Stop } from "../../__v3api";
import { TypedRoutes, RouteWithDirections } from "./__stop";
import RouteCard from "./RouteCard";

interface Props {
  routes: TypedRoutes[];
  stop: Stop;
}

const allRoutes = (typedRoutes: TypedRoutes[]): RouteWithDirections[] =>
  typedRoutes.reduce(
    (acc: RouteWithDirections[], typeAndRoutes: TypedRoutes) =>
      acc.concat(typeAndRoutes.routes),
    []
  );

const Departures = ({ routes, stop }: Props): ReactElement<HTMLElement> => (
  <>
    <h2>Departures</h2>

    <div className="m-stop-page__departures">
      {allRoutes(routes).map(routeWithDirections => (
        <RouteCard
          key={routeWithDirections.route.id}
          route={routeWithDirections.route}
          directions={routeWithDirections.directions}
          stop={stop}
        />
      ))}
    </div>
  </>
);

export default Departures;
