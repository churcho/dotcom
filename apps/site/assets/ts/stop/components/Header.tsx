import React, { ReactElement } from "react";
import { Stop, Route } from "../../v3api";
import TabComponent from "./Tab";
import { Tab, TypedRoutes } from "./__stop";
import {
  accessibleIcon,
  parkingIcon,
  redLineIcon,
  mattapanLineIcon,
  orangeLineIcon,
  blueLineIcon,
  greenLineIcon,
  greenBLineIcon,
  greenELineIcon,
  greenDLineIcon,
  greenCLineIcon,
  busIcon,
  commuterRailIcon
} from "../../helpers/icon";

interface Props {
  stop: Stop;
  routes: TypedRoutes[];
  tabs: Tab[];
  zoneNumber: string;
}

interface ModeIconType {
  [propName: string]: ReactElement<HTMLElement>;
}

const subwayModeIds = [
  "Blue",
  "Green",
  "Green-B",
  "Green-C",
  "Green-D",
  "Green-E",
  "Mattapan",
  "Orange",
  "Red"
];

const accessible = ({
  accessibility
}: Stop): ReactElement<HTMLElement> | false =>
  accessibility.includes("accessible") && (
    <a href="#accessibility" className="station__header-feature">
      <span className="station__header-icon">
        {accessibleIcon("c-svg__icon-accessible-default")}
      </span>
    </a>
  );

// eslint-disable-next-line typescript/camelcase
const parking = ({ parking_lots }: Stop): ReactElement<HTMLElement> | false =>
  // eslint-disable-next-line typescript/camelcase
  parking_lots.length > 0 && (
    <a href="#parking" className="station__header-feature">
      <span className="station__header-icon">
        {parkingIcon("c-svg__icon-parking-default")}
      </span>
    </a>
  );

const modeType = (modeId: string): string => {
  if (modeId.startsWith("CR-")) return "CR";

  if (subwayModeIds.includes(modeId)) return modeId;

  return "Bus";
};

const modeIcon = (modeId: string): JSX.Element | undefined => {
  if (modeId.startsWith("CR-"))
    return commuterRailIcon("c-svg__icon-commuter-rail-default");
  if (modeId === "Blue") return blueLineIcon("c-svg__icon-blue-line-default");
  if (modeId === "Green")
    return greenLineIcon("c-svg__icon-green-line-default");
  if (modeId === "Green-B")
    return greenBLineIcon("c-svg__icon-green-b-line-default");
  if (modeId === "Green-C")
    return greenCLineIcon("c-svg__icon-green-c-line-default");
  if (modeId === "Green-D")
    return greenDLineIcon("c-svg__icon-green-d-line-default");
  if (modeId === "Green-E")
    return greenELineIcon("c-svg__icon-green-e-line-default");
  if (modeId === "Mattapan")
    return mattapanLineIcon("c-svg__icon-mattapan-line-default");
  if (modeId === "Orange")
    return orangeLineIcon("c-svg__icon-orange-line-default");
  if (modeId === "Red") return redLineIcon("c-svg__icon-red-line-default");

  return busIcon("c-svg__icon-bus-default");
};

const modeLink = (modeId: string): string => {
  if (modeId.startsWith("CR-"))
    return "/stops/place-sstat?tab=departures#commuter-rail-schedule";
  if (subwayModeIds.includes(modeId))
    return "/stops/place-sstat?tab=departures#subway-schedule";

  return "/stops/place-sstat?tab=departures#bus-schedule";
};

const modeIconFeature = ({ id }: Route): ReactElement<HTMLElement> => (
  <a href={modeLink(id)} key={modeType(id)} className="station__header-feature">
    <span className="station__header-icon">{modeIcon(id)}</span>
  </a>
);

const iconableRoutesForType = ({
  // eslint-disable-next-line typescript/camelcase
  group_name,
  routes
}: TypedRoutes): Route[] => {
  // eslint-disable-next-line typescript/camelcase
  if (group_name === "subway") return routes;

  return [routes[0]];
};

const iconableRoutes = (typedRoutes: TypedRoutes[]): Route[] =>
  typedRoutes.reduce(
    (acc: Route[], typeAndRoutes: TypedRoutes) =>
      acc.concat(iconableRoutesForType(typeAndRoutes)),
    []
  );

const modes = (
  typedRoutes: TypedRoutes[]
): ReactElement<HTMLElement> | null => (
  <>{iconableRoutes(typedRoutes).map(route => modeIconFeature(route))}</>
);

const crZone = (zoneNumber: string): ReactElement<HTMLElement> | false =>
  !!zoneNumber &&
  zoneNumber.length > 0 && (
    <a
      href="/stops/place-sstat?tab=info#commuter-fares"
      className="station__header-feature"
    >
      <span className="station__header-icon c-icon__cr-zone">
        {`Zone ${zoneNumber}`}
      </span>
    </a>
  );

const features = (
  stop: Stop,
  routes: TypedRoutes[],
  zoneNumber: string
): ReactElement<HTMLElement> => (
  <div className="station__header-features">
    {modes(routes)}
    {crZone(zoneNumber)}
    {accessible(stop)}
    {parking(stop)}
  </div>
);

const Header = ({
  stop,
  routes,
  tabs,
  zoneNumber
}: Props): ReactElement<HTMLElement> => (
  <div className="station__header">
    <div className="station__header-container">
      <h1 className="station__name station__name--upcase">{stop.name}</h1>

      {features(stop, routes, zoneNumber)}

      <div className="header-tabs">
        {tabs.map(tab => (
          <TabComponent key={tab.id} tab={tab} />
        ))}
      </div>
    </div>
  </div>
);

export default Header;
