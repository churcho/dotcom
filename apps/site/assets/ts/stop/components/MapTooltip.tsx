import React, { ReactElement } from "react";
import { Stop, Route } from "../../__v3api";
import StopCard from "./StopCard";

interface Props {
  stop: Stop;
  routes: Route[];
}

const MapTooltip = ({ stop, routes }: Props): ReactElement<HTMLElement> => (
  <StopCard routes={routes} stop={stop} />
);

export default MapTooltip;
