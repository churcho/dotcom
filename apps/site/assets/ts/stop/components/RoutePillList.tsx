import React, { ReactElement } from "react";
import { Route } from "../../__v3api";
import { TypedRoutes } from "./__stop";
import { breakTextAtSlash } from "../../helpers/text";

interface RoutePillListProps {
  routes: TypedRoutes[];
}

interface RoutePillProps {
  route: Route;
}

const routeToClass = (name: string): string =>
  name.toLowerCase().replace(" ", "-");

const busName = (name: string): string =>
  name.startsWith("SL") ? "silver-line" : "bus";

const modeNameForBg = ({ name, type }: Route): string => {
  switch (type) {
    case 0:
    case 1:
      return routeToClass(name);
    case 2:
      return "commuter-rail";
    case 4:
      return "ferry";
    default:
      return busName(name);
  }
};

const modeBgClass = (route: Route): string => `u-bg--${modeNameForBg(route)}`;

const RoutePill = ({ route }: RoutePillProps): ReactElement<HTMLElement> => (
  <a
    href={`/schedules/${route.id}`}
    className={`
      m-stop-page__header-feature
      m-stop-page__header-description
      u-small-caps
      ${modeBgClass(route)}
    `}
  >
    {breakTextAtSlash(route.name)}
  </a>
);

const RoutePillList = ({
  routes
}: RoutePillListProps): ReactElement<HTMLElement> => (
  <div className="m-route-pills">
    {routes.map(typedRoute => (
      <div
        key={typedRoute.group_name}
        className={`m-route-pills--${typedRoute.group_name}`}
      >
        {typedRoute.routes.map(routeWithDirections => (
          <RoutePill
            key={routeWithDirections.route.id}
            route={routeWithDirections.route}
          />
        ))}
      </div>
    ))}
  </div>
);

export default RoutePillList;
