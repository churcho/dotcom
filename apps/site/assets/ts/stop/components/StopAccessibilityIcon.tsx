import React, { ReactElement } from "react";
import { Stop } from "../../__v3api";
import { accessibleIcon } from "../../helpers/icon";
import { Dispatch, clickFeaturePillAction } from "../state";

export default (
  { accessibility }: Stop,
  dispatch?: Dispatch
): ReactElement<HTMLElement> | false =>
  accessibility.includes("accessible") && (
    <a
      className="m-stop-page__access-icon"
      href="#header-accessibility"
      onClick={() =>
        dispatch && dispatch(clickFeaturePillAction("accessibility"))
      }
    >
      <span className="m-stop-page__icon">
        {accessibleIcon("c-svg__icon-accessible-default")}
      </span>
    </a>
  );
