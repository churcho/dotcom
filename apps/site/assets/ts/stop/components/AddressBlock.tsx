import React, { ReactElement } from "react";
import RoutePillList from "./RoutePillList";
import { Stop } from "../../__v3api";
import { TypedRoutes } from "./__stop";
import renderSvg from "../../helpers/render-svg";

// @ts-ignore
import streetViewSvg from "../../../static/images/icon-street-view-default.svg";

interface Props {
  routes: TypedRoutes[];
  stop: Stop;
  encoder?: (str: string) => string;
}

const renderAddress = (address: string): ReactElement<HTMLElement> => (
  <div className="m-stop-page__address">
    <h3 className="u-small-caps">Address</h3>
    <div className="h3">{address}</div>
  </div>
);

const latLngString = (stop: Stop): string =>
  `${stop.latitude},${stop.longitude}`;

const locationQuery = (stop: Stop, encoder?: (str: string) => string): string =>
  stop.address && encoder ? encoder(stop.address) : latLngString(stop);

const directionLink = (stop: Stop, encoder?: (str: string) => string): string =>
  `https://www.google.com/maps/dir/?api=1&destination=${locationQuery(
    stop,
    encoder
  )}`;

const streetViewLink = (stop: Stop): string =>
  `https://www.google.com/maps/@?api=1&map_action=pano&viewpoint=${latLngString(
    stop
  )}`;

const AddressBlock = ({
  routes,
  stop,
  encoder
}: Props): ReactElement<HTMLElement> => (
  <div className="m-stop-page__address-block">
    {stop.address && renderAddress(stop.address)}
    <div className="m-stop-page__address-links">
      <div className="m-stop-page__address-link">
        <a
          href={directionLink(stop, encoder)}
          className="btn btn-primary"
          target="_blank"
          rel="noopener noreferrer"
        >
          Get directions to this station
        </a>
      </div>
      <div className="m-stop-page__address-link">
        <a
          href={streetViewLink(stop)}
          target="_blank"
          rel="noopener noreferrer"
        >
          {renderSvg(
            "c-svg__icon-street-view m-stop-page__street-view-icon",
            streetViewSvg
          )}
          Street View
        </a>
      </div>
    </div>
    <RoutePillList routes={routes} />
  </div>
);

export default AddressBlock;
