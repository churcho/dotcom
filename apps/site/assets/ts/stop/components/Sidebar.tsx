import React, { ReactElement } from "react";
import ExpandableBlock from "../../app/ExpandableBlock";
import RoutePillList from "./RoutePillList";
// @ts-ignore
import accessibleIcon from "../../../static/images/icon-accessible-default.svg";
import ParkingInfo from "./ParkingInfo";
import { Stop } from "../../__v3api";
import { TypedRoutes } from "./__stop";

interface Props {
  stop: Stop;
  routes: TypedRoutes[];
}

const Sidebar = ({ stop, routes }: Props): ReactElement<HTMLElement> => (
  <div className="m-stop-page__sidebar">
    <RoutePillList routes={routes} />

    <h2>Features</h2>

    <ExpandableBlock
      initiallyExpanded={false}
      id="accessibility"
      header={{
        text: "Accessibility",
        iconSvgText: accessibleIcon
      }}
    >
      <div
        // eslint-disable-next-line react/no-danger
        dangerouslySetInnerHTML={{
          __html:
            "<p>South Station is accessible. It has the following features:</p><p>This is a test</p>"
        }}
      />
    </ExpandableBlock>

    <ParkingInfo stop={stop} />
  </div>
);

export default Sidebar;
