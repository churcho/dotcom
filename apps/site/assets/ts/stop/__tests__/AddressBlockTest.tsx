import React from "react";
import renderer from "react-test-renderer";
import stopData from "./stopData.json";
import { StopPageData } from "../components/__stop";
import AddressBlock from "../components/AddressBlock";

const data = JSON.parse(JSON.stringify(stopData)) as StopPageData;

it("renders", () => {
  document.body.innerHTML =
    '<div><div id="react-root"><div id="test"></div></div></div>';

  const tree = renderer.create(<AddressBlock routes={data.routes} />).toJSON();

  expect(tree).toMatchSnapshot();
});
