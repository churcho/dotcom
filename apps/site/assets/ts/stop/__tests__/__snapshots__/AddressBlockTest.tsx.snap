// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="m-stop-page__address-block"
>
  <div
    className="m-stop-page__address"
  >
    <h3
      className="u-small-caps"
    >
      Address
    </h3>
    <div
      className="h3"
    >
      700 Atlantic Ave, Boston, MA 02110
    </div>
  </div>
  <div
    className="m-stop-page__address-links"
  >
    <div
      className="m-stop-page__address-link"
    >
      <a
        className="btn btn-primary"
        href="https://www.google.com/maps/dir/?api=1&destination=700%20Atlantic%20Ave%2C%20Boston%2C%20MA%2002110"
        rel="noopener noreferrer"
        target="_blank"
      >
        Get directions to this station
      </a>
    </div>
    <div
      className="m-stop-page__address-link"
    >
      <a
        href="https://www.google.com/maps/@?api=1&map_action=pano&viewpoint=42.352271,-71.055242"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span
          aria-hidden="true"
          className="notranslate c-svg__icon-street-view m-stop-page__street-view-icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
        Street View
      </a>
    </div>
  </div>
  <div
    className="m-route-pills"
  >
    <div
      className="m-route-pills--subway"
    >
      <div
        className="
      m-stop-page__header-feature
      m-stop-page__header-description
      u-small-caps
      u-bg--red-line
    "
      >
        Red Line
      </div>
    </div>
    <div
      className="m-route-pills--bus"
    >
      <div
        className="
      m-stop-page__header-feature
      m-stop-page__header-description
      u-small-caps
      u-bg--silver-line
    "
      >
        SL1
      </div>
      <div
        className="
      m-stop-page__header-feature
      m-stop-page__header-description
      u-small-caps
      u-bg--silver-line
    "
      >
        SL2
      </div>
      <div
        className="
      m-stop-page__header-feature
      m-stop-page__header-description
      u-small-caps
      u-bg--silver-line
    "
      >
        SL3
      </div>
    </div>
    <div
      className="m-route-pills--commuter_rail"
    >
      <div
        className="
      m-stop-page__header-feature
      m-stop-page__header-description
      u-small-caps
      u-bg--commuter-rail
    "
      >
        Fairmount Line
      </div>
      <div
        className="
      m-stop-page__header-feature
      m-stop-page__header-description
      u-small-caps
      u-bg--commuter-rail
    "
      >
        Framingham/​Worcester Line
      </div>
      <div
        className="
      m-stop-page__header-feature
      m-stop-page__header-description
      u-small-caps
      u-bg--commuter-rail
    "
      >
        Franklin Line
      </div>
      <div
        className="
      m-stop-page__header-feature
      m-stop-page__header-description
      u-small-caps
      u-bg--commuter-rail
    "
      >
        Greenbush Line
      </div>
      <div
        className="
      m-stop-page__header-feature
      m-stop-page__header-description
      u-small-caps
      u-bg--commuter-rail
    "
      >
        Kingston/​Plymouth Line
      </div>
      <div
        className="
      m-stop-page__header-feature
      m-stop-page__header-description
      u-small-caps
      u-bg--commuter-rail
    "
      >
        Middleborough/​Lakeville Line
      </div>
      <div
        className="
      m-stop-page__header-feature
      m-stop-page__header-description
      u-small-caps
      u-bg--commuter-rail
    "
      >
        Needham Line
      </div>
      <div
        className="
      m-stop-page__header-feature
      m-stop-page__header-description
      u-small-caps
      u-bg--commuter-rail
    "
      >
        Providence/​Stoughton Line
      </div>
    </div>
  </div>
</div>
`;
